vault/identity_store_entities.go

```diff
@@ -0,0 +1,463 @@
+		}
+
+		if fromEntity.ID != lockFromEntity.ID {
+			return logical.ErrorResponse("acquired lock for an undesired entity"), nil
+		}
+
+		for _, identity := range fromEntity.Identities {
+			// Special case the identities that gets transfered over due to merge
+			// operation. This might also aid in controlling any actions to be
+			// taken on the merged identities.
+			identity.MountType = "EntityAlias"
+
+			// Set the desired entity id
+			identity.EntityID = toEntity.ID
+
+			// Set the entity id of which this identity is now an alias to
+			identity.MergedFrom = fromEntity.ID
+
+			err = i.memDBUpsertIdentityInTxn(txn, identity)
+			if err != nil {
+				return nil, fmt.Errorf("failed to update identity during merge: %v", err)
+			}
+
+			// Add the identity to the desired entity
+			toEntity.Identities = append(toEntity.Identities, identity)
+		}
+
+		// Add the entity from which we are merging from to the list of entities
+		// the entity we are merging into is composed of
+		toEntity.MergedEntities = append(toEntity.MergedEntities, fromEntity.ID)
+
+		// Delete the entity which we are merging from in MemDB using the same transaction
+		err = i.memDBDeleteEntityInTxn(txn, fromEntity)
+		if err != nil {
+			return nil, err
+		}
+
+		// Delete the entity which we are merging from in storage
+		err = i.storagePacker.DeleteItem(fromEntity.ID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Persist the entity which we are merging to
+	err = i.storagePacker.PutItem(toEntity)
+	if err != nil {
+		return nil, err
+	}
+
+	// Committing the transaction *after* successfully performing storage
+	// persistence
+	txn.Commit()
+
+	return nil, nil
+}
+
+// pathEntityRegister is used to register a new entity
+func (i *identityStore) pathEntityRegister(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	return i.handleEntityUpdateCommon(req, d, nil)
+}
+
+// pathEntityIDUpdate is used to update an entity based on the given entity ID
+func (i *identityStore) pathEntityIDUpdate(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	// Get entity id
+	entityID := d.Get("id").(string)
+
+	if entityID == "" {
+		return logical.ErrorResponse("missing entity id"), nil
+	}
+
+	entity, err := i.memDBEntityByID(entityID)
+	if err != nil {
+		return nil, err
+	}
+	if entity == nil {
+		return nil, fmt.Errorf("invalid entity id")
+	}
+
+	return i.handleEntityUpdateCommon(req, d, entity)
+}
+
+// handleEntityUpdateCommon is used to update an entity
+func (i *identityStore) handleEntityUpdateCommon(req *logical.Request, d *framework.FieldData, entity *entityStorageEntry) (*logical.Response, error) {
+	var err error
+	var newEntity bool
+
+	// Entity will be nil when a new entity is being registered; create a new
+	// struct in that case.
+	if entity == nil {
+		entity = &entityStorageEntry{}
+		newEntity = true
+	}
+
+	// Update the policies
+	entity.Policies = d.Get("policies").([]string)
```

Comment:
I think this should be `GetOk` so that entities can be updated without having to specify policy value every time.